adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -121,6 +121,121 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
     replaceRdd(tFn(rdd))
   }
 
+  private[rdd] val sortedTrait: SortedTrait
+  private[rdd] def sorted = sortedTrait.sorted
+  private[rdd] def partitionMap = sortedTrait.partitionMap
+
+  /**
+   * This case class is the source of all sorted knowledge handling in ADAM. We
+   * have it as an inner class currently because there were some access issues
+   */
+  private[rdd] case class SortedTrait(sorted: Boolean,
+                                      partitionMap: Option[Seq[(ReferenceRegion, ReferenceRegion)]]) extends Serializable
+  /**
+   * This repartition method repartitions all data in this.rdd and distributes it as evenly as possible
+   * into the number of partitions provided. The data in this.rdd is already sorted (by ReferenceRegion.start)
+   * so we ensure that it is kept that way as we move data around.
+   *
+   * Data is repartitioned in lists to maintain the sorted order. After the repartition, the lists are placed
+   * in the correct order based on their original sorted order.
+   *
+   * @param partitions the number of partitions to repartition this.rdd into
+   * @return a new SortedGenomicRDDMixIn with the RDD partitioned evenly
+   */
+  private def evenlyRepartition(partitions: Int)(implicit tTag: ClassTag[T]): GenomicRDD[T, U] = {
+    assert(sorted)
```

Comment:
I don't have an opinion about the assertion, but if we have an assertion that is not obvious (e.g., something like https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala#L175 --> "fields should not be null" would be an obvious assert/require) then we should document why that assertion is important inline. Or, even better, we should document it through the message that the assert throws, so that when it invariably shows up in a stack trace, it isn't just stack barf.